pkg/controllers/externalsecret/externalsecret_controller.go

```diff
@@ -517,57 +530,61 @@ func (r *Reconciler) getStore(ctx context.Context, externalSecret *esv1beta1.Ext
 }
 
 // getProviderSecretData returns the provider's secret data with the provided ExternalSecret.
-func (r *Reconciler) getProviderSecretData(ctx context.Context, providerClient esv1beta1.SecretsClient, externalSecret *esv1beta1.ExternalSecret) (map[string][]byte, error) {
+func (r *Reconciler) getProviderSecretData(
+	ctx context.Context, providerClient esv1beta1.SecretsClient, externalSecret *esv1beta1.ExternalSecret) (map[string][]byte, esv1beta1.SecretsMetadata, error) {
 	providerData := make(map[string][]byte)
+	var providerSecretMetadata esv1beta1.SecretsMetadata
 
 	for i, remoteRef := range externalSecret.Spec.DataFrom {
 		var secretMap map[string][]byte
 		var err error
 		if remoteRef.Find != nil {
-			secretMap, err = providerClient.GetAllSecrets(ctx, *remoteRef.Find)
+			secretMap, providerSecretMetadata, err = providerClient.GetAllSecrets(ctx, *remoteRef.Find)
 			if errors.Is(err, esv1beta1.NoSecretErr) && externalSecret.Spec.Target.DeletionPolicy != esv1beta1.DeletionPolicyRetain {
 				r.recorder.Event(externalSecret, v1.EventTypeNormal, esv1beta1.ReasonDeleted, fmt.Sprintf("secret does not exist at provider using .dataFrom[%d]", i))
 				continue
 			}
 			if err != nil {
-				return nil, err
+				return nil, esv1beta1.SecretsMetadata{}, err
 			}
 			secretMap, err = utils.ConvertKeys(remoteRef.Find.ConversionStrategy, secretMap)
 			if err != nil {
-				return nil, fmt.Errorf(errConvert, err)
+				return nil, esv1beta1.SecretsMetadata{}, fmt.Errorf(errConvert, err)
 			}
 		} else if remoteRef.Extract != nil {
-			secretMap, err = providerClient.GetSecretMap(ctx, *remoteRef.Extract)
+			secretMap, providerSecretMetadata, err = providerClient.GetSecretMap(ctx, *remoteRef.Extract)
 			if errors.Is(err, esv1beta1.NoSecretErr) && externalSecret.Spec.Target.DeletionPolicy != esv1beta1.DeletionPolicyRetain {
 				r.recorder.Event(externalSecret, v1.EventTypeNormal, esv1beta1.ReasonDeleted, fmt.Sprintf("secret does not exist at provider using .dataFrom[%d]", i))
 				continue
 			}
 			if err != nil {
-				return nil, err
+				return nil, esv1beta1.SecretsMetadata{}, err
 			}
 			secretMap, err = utils.ConvertKeys(remoteRef.Extract.ConversionStrategy, secretMap)
 			if err != nil {
-				return nil, fmt.Errorf(errConvert, err)
+				return nil, esv1beta1.SecretsMetadata{}, fmt.Errorf(errConvert, err)
 			}
 		}
 
 		providerData = utils.MergeByteMap(providerData, secretMap)
 	}
 
 	for i, secretRef := range externalSecret.Spec.Data {
-		secretData, err := providerClient.GetSecret(ctx, secretRef.RemoteRef)
+		// SecretsMetadata is ignored here when we call the provider multiple times
+		// We don't know what to do if multiple metadatas with differents values are returned
+		secretData, _, err := providerClient.GetSecret(ctx, secretRef.RemoteRef)
```

Comment:
This problem would also happen with `dataFrom` operations (the for loop in line 538).

Reply:
I've handle this case also. However I'm thinking what we should do in this case
- If we call the GetSecrets* method multiple times, we ignore it
- Or should we try to be smarter, and maybe take the smallest leaseTimeout returned by all the calls 

What do you think is best ?

Reply:
I think in both cases we would probably need a smarter approach..

 IMO what I think it's desireable from a Provider call perspective is to only call the bits of the secrets which actually had their Lease Timeouts expired, and not everything else (this is effectively a cache timeout per secret Keys, while the refresh interval is a cache timeout for the secret as a whole). So I think maybe we need to add the LeaseTimeout information as a map to the original 'data/dataFrom' blocks, and check for it at the `for` loop level.

WDYT?

Reply:
Not sure the map is relevant in all cases. If I take an example with the AWS Engine itself, the problem is that each call to the provider generates a new pair of access_key/secret_key. This pair is atomic, you cannot use the access_key generated for one secret with the secret_key generated for another secret, making some usages of external secrets irrelevant. For instance, we cannot do this, because we would have two values coming from different pairs of secrets

```yml
data:
  - secretKey: access_key
    remoteRef:
      key: /vault-role-name
  - secretKey: secret_key
    remoteRef:
      key: /vault-role-name
```

This is why in my example I use dataFrom to have only a single call to the provider

```yml
  dataFrom:
  - extract:
      conversionStrategy: Default
      key: role-name  # Role name in the Vault AWS Engine
...    
     template:
      data:
        AWS_ACCESS_KEY_ID: '{{ .access_key }}'
        AWS_SECRET_ACCESS_KEY: '{{ .secret_key }}'
```

I think that the code is not designed to handle "ephemeral" secrets and that we are just trying to work around that 😅

Reply:
I'm more trying to extend the concept outside of this use case in particular. Generally speaking if we have a given provider `ABC` that implements Lease timeout, ideally we want to keep track of all the operations the returned it, hence the map. I also believe we should implement an error for the `GetSecret` when using AWS Engine, as I believe it is not useful at all to have only half the keys - now that you mentioned it :).

Also, some general questions to AWS engine, as I am not familiar with. Whenever we call the endpoint we are generating new creds. Isn't there anyway to consume this creds while they are still valid, without generating new ones?

Reply:
I also believe that maybe having a CRD to generate the Secret, and then having something just consuming whatever was generated, would work much better. Specially if there are parameters that we can configure per provider (for instance, the lease timeout?).

Reply:
I'm not sure what you imagine about the CRD, could you share a quick example ?

Reply:
@julienvey It would be a CRD specifically designed to handle Secret generation...

@moolen @knelasevero WDYT? It seems to me that there are two things hapenning in the provider. The first one is a Secrets Generation, while the other one is the ExternalSecrets per se. Do you think it is worth to properly separate it?

> Also, some general questions to AWS engine, as I am not familiar with. Whenever we call the endpoint we are generating new creds. Isn't there anyway to consume this creds while they are still valid, without generating new ones? If not, I believe we should implement an error for the GetSecret when using AWS Engine, as I believe it is not useful at all to have only half the keys - now that you mentioned it :).

@julienvey can you also help out with this question? It is still not clear for me :)

Reply:
@gusfcarvalho Sure :)

- Anytime you call the endpoint to generate new credentials, you will get a pair of secret_access_key/access_key_id, and also a leaseID (vault specific)
- Once you get the pair of credentials, you have to store it somewhere because there is no way to read them back
- The credentials validity can be defined in the vault secret engine itself (default value), or you can customize it for each request to the engine (as long as it is a shorter duration than the default one).
- To know if the credentials are still valid, you can use the leaseID, and ask vault about it at any time. There's an endpoint to tell you when the lease will expire (and the credentials will be deactivated)

If you want to get in more details, there are several "ways" vault can generate a secret, some have caveats (with iam_user, the generation is "eventually consistent", you have a delay of 5 to 10s between the secret generation, and when you can actually use it)

The docs for more details : https://www.vaultproject.io/docs/secrets/aws

Reply:
Thanks for the feeback @moolen 

I've started to rewrite my code as you describe (have a struct `secretEngine` that can handle config for `secretEngine.aws` or `secretEngine.kv`)

However, just to be clear, this will break existing configurations. Is it ok for you ? If haven't found an "Upgrading guide" in the docs, but I can work on this if needed

Reply:
> this will break existing configurations

This shouldn't happen. We have a [stability/support](https://external-secrets.io/v0.5.8/stability-support/) guarantee that we want to follow. I.e. changes must be compatible.